Data specification versioning

CristianGimenez opened the pull request:
As mentioned in the ticket, we have agreed not to implement the finalisation feature mentioned origianally in the issue. In this implementation you will find the following features

- Creating a new version of a finalised (submitted) data specification.
- When listing your data specifications, or shared data specifications now, only the latest model version is returned.
- On the details of a data specification there is a dropdown that can be used to see a different version.

Since there is no way to finalise the data specifications, an instance of mdm-ui is required to run manual tests, so we need to finalise manually data specifications to see the "create new version" button and to see how they are now listed or not.

Also the version selector component infers the current "draft" version to be the next available major version.

Draft until: [mdm-resources/106](https://github.com/MauroDataMapper/mdm-resources/pull/107) and [plugin-explorer/34](https://github.com/MauroDataMapper-Plugins/mdm-plugin-explorer/issues/34) are resolved

Closes #315

CristianGimenez committed:
gh-315: create new version and version controls need styles

CristianGimenez committed:
gh-315: styles and new endpoint connection

CristianGimenez committed:
gh-315: adding tests and eslint fixes

NigelPalmer commented on src/app/mauro/data-model.service.ts:
```diff
@@ -368,7 +370,7 @@ export class DataModelService {
 
     return this.endpoints.dataModel
-      .newBranchModelVersion(model.id, {})
+      .newBranchModelVersion(model.id, { asynchronous: false })
```
Why have we turned off asynchronus here?

CristianGimenez replied:
To be aligned with mdm-ui, changed back

NigelPalmer commented on src/styles/_versioncontrol.scss:
```diff
@@ -0,0 +1,32 @@
+*/
+.version_control {
+  .select-label {
+    color: var(--theme-color-primary-contrast) !important;
```
This should be `$color-primary-contrast` . You should not be calling var(..) directly.

CristianGimenez replied:
done

NigelPalmer commented on src/styles/_versioncontrol.scss:
```diff
@@ -0,0 +1,32 @@
+    color: var(--theme-color-primary-contrast) !important;
+  }
+  .mat-form-field-outline {
+    color: var(--theme-color-primary-contrast) !important;
```
This should be $color-primary-contrast . You should not be calling var(..) directly.

CristianGimenez replied:
done

NigelPalmer commented on src/styles/_versioncontrol.scss:
```diff
@@ -0,0 +1,32 @@
+    color: var(--theme-color-primary-contrast) !important;
+  }
+  .mat-select-value {
+    color: var(--theme-color-primary-contrast) !important;
```
This should be $color-primary-contrast . You should not be calling var(..) directly.

CristianGimenez replied:
done

NigelPalmer commented on src/styles/_versioncontrol.scss:
```diff
@@ -0,0 +1,32 @@
+    color: var(--theme-color-primary-contrast) !important;
+  }
+  .mat-select-arrow {
+    color: var(--theme-color-primary-contrast) !important;
```
This should be $color-primary-contrast . You should not be calling var(..) directly.

CristianGimenez replied:
done

NigelPalmer reviewed (changes requested):
Overall good job, and the main functionality is there. We do have some problems though, that need to be fixed.

- [ ] I don't like the number 1 just floating on the header bar of a Data Specification. It has no context or meaning to the user and simple looks odd. Once submitted the version number changed to V1.0.0. It should be displayed like this from the beginning.  **- Please fix.**
- [X] Copied a specification and then deletes an element from the copy. The original kept that element. _No work required_
- [ ] I added a cohort query and data query to V2.0.0. When I change the version to be V1.0.0 the queries were still displayed. The queries should be associated with the version of the specification. (Or possibly this is the next couple of errors manifesting, read on...) **- Please fix.**
- [ ] Once I had V1.0.0, V2.0.0 and unsubmitted v3.0.0, if I changed the Cohort query for v3.0.0, then the same changes were applied to v1.0.0 (but NOT v2.0.0). Changing the Cohort Query should only affect the latest version. **- Please fix.**
- [ ] Same happened if I updated and saved the latest Data Query. The changes were also applied to V1.0.0. **- Please fix.**
- [ ] You can click "New Version" at any time. If you already have an unsubmitted version then you will get the error `Http failure response for http://localhost:8080/api/dataModels/9a93d3f1-ddd8-48e0-895a-6b577f0d3f5e/newBranchModelVersion: 403 OK`. The "New Version" option should only be available if ALL versions are submitted. I suggest disabling the button rather than hiding it. Once I get this error, I am unable to create new versions for any Data Specification until I log out and back in again. **- Please fix.**

CristianGimenez committed:
gh-315: Pr comments 1

- new version is now disabled if a new version already exists
- new version is now asynchronous
- fixed an issue where unsubmitted data specs got a "1" rather than "V1.0.0"

CristianGimenez commented:
> Overall good job, and the main functionality is there. We do have some problems though, that need to be fixed.
> 
> * [ ]  I don't like the number 1 just floating on the header bar of a Data Specification. It has no context or meaning to the user and simple looks odd. Once submitted the version number changed to V1.0.0. It should be displayed like this from the beginning.  **- Please fix.**
> * [x]  Copied a specification and then deletes an element from the copy. The original kept that element. _No work required_
> * [ ]  I added a cohort query and data query to V2.0.0. When I change the version to be V1.0.0 the queries were still displayed. The queries should be associated with the version of the specification. (Or possibly this is the next couple of errors manifesting, read on...) **- Please fix.**
> * [ ]  Once I had V1.0.0, V2.0.0 and unsubmitted v3.0.0, if I changed the Cohort query for v3.0.0, then the same changes were applied to v1.0.0 (but NOT v2.0.0). Changing the Cohort Query should only affect the latest version. **- Please fix.**
> * [ ]  Same happened if I updated and saved the latest Data Query. The changes were also applied to V1.0.0. **- Please fix.**
> * [ ]  You can click "New Version" at any time. If you already have an unsubmitted version then you will get the error `Http failure response for http://localhost:8080/api/dataModels/9a93d3f1-ddd8-48e0-895a-6b577f0d3f5e/newBranchModelVersion: 403 OK`. The "New Version" option should only be available if ALL versions are submitted. I suggest disabling the button rather than hiding it. Once I get this error, I am unable to create new versions for any Data Specification until I log out and back in again. **- Please fix.**

I'm letting you check or uncheck those elements after you retest. I have just pushed a change fixing all the issues mentioned there but the ones about cohort and data queries which I'm investigating now.
I dont like how the disabled button looks like (and I think that is not a good UX practice to disable buttons without a clear indication of why are they disabled) but take a look and let me know what would you prefer. The button will be hidden if the dataspec is not submitted still.

CristianGimenez committed:
gh-315: fixing last issues from PRcomments

CristianGimenez committed:
Merge branch 'develop' into enhancement/gh-315

# Conflicts:
#	src/app/data-explorer/data-explorer.module.ts
#	src/app/data-explorer/data-specification-row/data-specification-row.component.html
#	src/app/data-explorer/data-specification-row/data-specification-row.component.ts
#	src/app/mauro/data-model.service.ts
#	src/app/mauro/plugins/plugin-research.resource.ts
#	src/app/mauro/research-plugin.service.ts
#	src/app/pages/my-data-specification-detail/my-data-specification-detail.component.html
#	src/app/pages/my-data-specification-detail/my-data-specification-detail.component.ts
#	src/app/pages/my-data-specifications/my-data-specifications.component.ts
#	src/app/pages/template-data-specifications/template-data-specifications.component.html

CristianGimenez committed:
gh-315: fix eslint after merge

CristianGimenez commented:
@NigelPalmer I have fixed all your the issues you raised and merge conflicts, so if you are happy to retest, this would be ready to merge now.

NigelPalmer reviewed (changes requested):
- [X] V1.0.0 now displayed in all cases on the Data Specification list.
- [X] Adding Cohort and Data queries only adds them to the latest version.

- [ ] The disabled button is so dark it is barely visible. We either need to disable it with a brighter colour or hide it.
- [ ] After I submit a specification, the "New Version" button does not activate until you click away and come back.
- [ ] When I try to create a new version after submitting I get a 403 error. Steps to reproduce:
1. Start a clean browser and go to MDE 
2. Log in.
3. Select a data specification and submit.
4. Click "Back to my Data Specifications"
5. Click on the same data specification
6. Click on "New Version"
7. Click on "Yes"
I can only clear this problem by closing the browser and reopening it. This makes testing a burden since only submitted Data Specifications can have new versions created so I am having to constantly close and open my browser to be able to test this work. (Found another way around, if I submit and then share with the community, then I am able to click back and then select the Data Specification again. Now when I click "New Version" no errors occur.)
- [ ] The version drop down is showing the selected version twice. (Or possibly it's a symptom of the next issue.)
![DuplicatedVersionNumber](https://github.com/MauroDataMapper/mdm-explorer/assets/110250803/83c7b49a-eaf6-471c-9ddd-6baeff77a697)
- [ ] I had 4 versions of the same Data Specification. I selected V2 and clicked on "New Version". This created a new v3, so now I have two v3s.
![CreatedNewVersionPlusOne](https://github.com/MauroDataMapper/mdm-explorer/assets/110250803/a04eb143-a409-4f7f-911d-25f4435d9d1d)
- [ ] I created a brand new Data Specification. This resulted in two version 1s.
![DuplicatedOnes](https://github.com/MauroDataMapper/mdm-explorer/assets/110250803/f1a28b6b-c1e4-4139-bda3-4edb50560f6f)

CristianGimenez committed:
gh315: bugfixing as per PR comments

CristianGimenez committed:
gh-315: fixing submit error and refactor component

CristianGimenez commented:
@NigelPalmer I have pushed a new commit with the following fixes/changes:
- New treeList method in the folders endpoint so we can replicate the "refresh" of mdm-ui, solving the submit problems.
- Refactored my-data-specification-detail.component.ts to remove the nested subscriptions that I've found.
- Changed the styles of the new version button, hopefully is more visible now, I've also added a tooltip (which requires a bit more of boiler plate since it does not work well in disabled buttons)
- Fixed bugs mentioned as duplicated versions etc.

Hopefully this is now working well and can be merged, could you retest and let me know any further changes required?

CristianGimenez committed:
gh-315: fixing eslint and tests

NigelPalmer commented on src/styles.scss:
```diff
@@ -37,3 +37,4 @@ Import more theme SASS files here to make them available to the app
 @import "styles/home";
 @import "styles/querybuilder";
 @import "styles/versioncontrol";
+@import "styles/_dataspecificationsactions";
```
I don't think the proceeding _ should be here since it is not present on any of the other references.

CristianGimenez replied:
You are right, alas I'm deleting that file anyway since it is no longer needed (it was all about the disabled button styles)

NigelPalmer commented on src/app/pages/my-data-specification-detail/my-data-specification-detail.component.ts:
```diff
@@ -203,12 +203,13 @@ export class MyDataSpecificationDetailComponent implements OnInit, OnDestroy {
         // Refresh the current state of the data specification in view
-        this.dataSpecification = mapToDataSpecification(dataModel);
+        this.setDataSpecification(mapToDataSpecification(dataModel));
+        // this.dataSpecification = mapToDataSpecification(dataModel);
```
Comment should be deleted

CristianGimenez replied:
deleted in the latest version

NigelPalmer reviewed (changes requested):
- [X] New Version button is now working correctly.
- [X] No errors are occuring.
- [ ] Tests are still failing in Jenkins?
- [ ] Please hide the "New Version" button instead of disabling it when it is not required.  Apologies for messing you about with this, but having tested it on this iteration, it has become clear to me that hiding the button makes a better user experience.

CristianGimenez committed:
gh-315: PR comments

disabling new version button
removing unneeded styles

NigelPalmer reviewed (approved):
Button now removed. I am not completely convinced by the version dropdown box jumping about, but I think it makes for a better user experience not having the button visible when it is not required. We can review this when creating the request management pages. Approved.
